File: app/schemas/ap_policies.py
"""Public API schemas for editable AP policies and their history."""

from datetime import datetime
from typing import Annotated, Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    StrictBool,
    StrictFloat,
    StrictInt,
    StrictStr,
    StringConstraints,
)

PolicyScalar = StrictStr | StrictInt | StrictFloat | StrictBool
PolicyNote = Annotated[str, StringConstraints(strip_whitespace=True, max_length=1000)]


class APPolicyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    key: str
    name: str
    description: str
    value_type: Literal["number", "enum", "boolean", "date"]
    value: PolicyScalar
    options: list[PolicyScalar] | None = None
    unit: str | None = None
    severity: Literal["block", "escalate", "advise"]
    active: bool
    version: int
    updated_at: datetime | None = None
    updated_by: str | None = None


class APPolicyListResponse(BaseModel):
    items: list[APPolicyResponse]
    total: int
    snapshot_label: str


class APPolicyUpdateRequest(BaseModel):
    value: PolicyScalar
    note: PolicyNote = ""


class APPolicyVersionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    version: int
    value: PolicyScalar
    previous_value: PolicyScalar | None = None
    changed_by: str | None = None
    changed_at: datetime
    note: str | None = None


class APPolicyHistoryResponse(BaseModel):
    policy_key: str
    items: list[APPolicyVersionResponse]
    total: int

File: frontend/src/types/ap-policies.ts
export type APPolicyValueType = 'number' | 'enum' | 'boolean' | 'date'
export type APPolicyValue = string | number | boolean
export type APPolicySeverity = 'block' | 'escalate' | 'advise'

export interface APPolicyBase {
  key: string
  name: string
  description: string
  options: APPolicyValue[] | null
  unit: string | null
  severity: APPolicySeverity
  active: boolean
  version: number
  updated_at: string | null
  updated_by: string | null
}

export interface APNumberPolicy extends APPolicyBase {
  value_type: 'number'
  value: number
}

export interface APEnumPolicy extends APPolicyBase {
  value_type: 'enum'
  value: string
}

export interface APBooleanPolicy extends APPolicyBase {
  value_type: 'boolean'
  value: boolean
}

export interface APDatePolicy extends APPolicyBase {
  value_type: 'date'
  value: string
}

export type APPolicy =
  | APNumberPolicy
  | APEnumPolicy
  | APBooleanPolicy
  | APDatePolicy

export interface APPolicyListResponse {
  items: APPolicy[]
  total: number
  snapshot_label: string
}

export interface APPolicyUpdateRequest {
  value: APPolicyValue
  note: string
}

export interface APPolicyVersion {
  version: number
  value: APPolicyValue
  previous_value: APPolicyValue | null
  changed_by: string | null
  changed_at: string
  note: string | null
}

export interface APPolicyHistoryResponse {
  policy_key: string
  items: APPolicyVersion[]
  total: number
}

/**
 * One policy, applied to one invoice, on one run.
 *
 * The backend writes a row per policy per run whether or not it fired — that is
 * what makes the gate auditable rather than merely logged. It reaches the
 * Workbench inside the item's untyped `context`, so treat it as unvalidated JSON
 * until it has been through `parsePolicyEvaluations`.
 */
export interface APPolicyEvaluation {
  policy_key: string
  policy_version: number
  threshold_value: unknown
  observed_value: unknown
  fired: boolean
  outcome: string | null
  explanation: string | null
}

export type APPolicyValidationTarget = Pick<APPolicy, 'value_type' | 'options'>

export type APPolicyValueValidationResult =
  | { valid: true; value: APPolicyValue }
  | { valid: false; error: string }
